client.go

```diff
@@ -0,0 +1,322 @@
+package influxdb
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/influxdata/influxdb-client-go/internal/gzip"
+	lp "github.com/influxdata/line-protocol"
+)
+
+const defaultMaxWait = 10 * time.Second
+
+// Client is a client for writing to influx.
+type Client struct {
+	httpClient       *http.Client
+	contentEncoding  string
+	compressionLevel int
+	url              *url.URL
+	password         string
+	username         string
+	l                sync.Mutex
+	maxRetries       int
+	errOnFieldErr    bool
+	userAgent        string
+	authorization    string
+	maxLineBytes     int
+}
+
+// New creates a new Client.  If httpClient is nil, it will use an http client with sane defaults.
+// The client is concurrency safe, so feel free to use it and abuse it to your heart's content.
+func New(httpClient *http.Client, options ...Option) (*Client, error) {
+	c := &Client{
+		httpClient:       httpClient,
+		contentEncoding:  "gzip",
+		compressionLevel: 4,
+		errOnFieldErr:    true,
+	}
+	if c.httpClient == nil {
+		c.httpClient = defaultHTTPClient()
+	}
+	c.url, _ = url.Parse(`http://127.0.0.1:9999/api/v2`)
+	c.userAgent = ua()
+	for i := range options {
+		if err := options[i](c); err != nil {
+			return nil, err
+		}
+	}
+	if c.authorization == "" && !(c.username != "" || c.password != "") {
+		return nil, errors.New("a token or a username and password is required, use WithToken(\"the_token\"), or use WithUserAndPass(\"the_username\",\"the_password\")")
+	}
+	return c, nil
+}
+
+// Ping checks the status of cluster.
+func (c *Client) Ping(ctx context.Context) (time.Duration, string, error) {
+	ts := time.Now()
+	req, err := http.NewRequest("GET", c.url.String()+"/ready", nil)
+	if err != nil {
+		return 0, "", err
+	}
+
+	req = req.WithContext(ctx)
+	resp, err := c.httpClient.Do(req)
+	dur := time.Since(ts)
+	if err != nil {
+		return dur, "", err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	dur = time.Since(ts)
+	if err != nil {
+		return dur, "", err
+	}
+
+	// we shouldn't see this
+	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
+		var err = errors.New(string(body))
+		return 0, "", err
+	}
+
+	version := resp.Header.Get("X-Influxdb-Version")
+	return dur, version, nil
+}
+
+// Write writes metrics to a bucket, and org.  It retries intelligently.
+// If the write is too big, it retries again, after breaing the payloads into two requests.
+func (c *Client) Write(ctx context.Context, bucket, org string, m ...Metric) (err error) {
+	if c.authorization == "" {
+		return errors.New("a token is requred for a write")
+	}
+	tries := uint64(0)
+	return c.write(ctx, bucket, org, &tries, m...)
+}
+
+func (c *Client) write(ctx context.Context, bucket, org string, triesPtr *uint64, m ...Metric) error {
+	//r, w := io.Pipe()
+	buf := &bytes.Buffer{}
+	e := lp.NewEncoder(buf)
+doRequest:
```

Comment:
I would love to see a refactor of this to remove the `goto` dependency. I'm not convinced we need that here when a loop could suffice and clarify things dramatically.

Reply:
I prefer the goto for this use case, its clearer almost every case involves breaking out of the loop.  The goto makes it clear that it is the unusual case where we retry.